### PR TITLE
compiler_fence documentation: emphasize synchronization, not reordering

### LINKED ISSUE #54962: compiler_fence SeqCst docs are incorrect
The documentation for `compiler_fence` with `SeqCst` say "no re-ordering of reads and writes across this point is allowed". As I understand it, `SeqCst` for a single thread doesn't give anything more than `AcqRel`. E.g., a store before the fence could be moved past a load after then fence. Is that correct?

cc @jonhoo @RalfJung

### LINKED ISSUE #129189: compiler_fence documentation is misleading
The current docs for [`compiler_fence`](https://doc.rust-lang.org/nightly/std/sync/atomic/fn.compiler_fence.html) explain that function in terms of reorderings:

> compiler_fence does not emit any machine code, but restricts the kinds of memory re-ordering the compiler is allowed to do. Specifically, depending on the given [Ordering](https://doc.rust-lang.org/nightly/std/sync/atomic/enum.Ordering.html) semantics, the compiler may be disallowed from moving reads or writes from before or after the call to the other side of the call to compiler_fence. Note that it does not prevent the hardware from doing such re-ordering. This is not a problem in a single-threaded, execution context, but when other threads may modify memory at the same time, stronger synchronization primitives such as [fence](https://doc.rust-lang.org/nightly/std/sync/atomic/fn.fence.html) are required.

That's somewhat misleading. For good reasons, our [`fence`](https://doc.rust-lang.org/nightly/std/sync/atomic/fn.fence.html) docs quickly transition from "reordering" to "synchronization" terminology:

> Depending on the specified order, a fence prevents the compiler and CPU from reordering certain types of memory operations around it. That creates synchronizes-with relationships between it and atomic operations or fences in other threads.

`compiler_fence` should also be explained in terms of synchronizes-with, to avoid raising any false expectations. Specifically, compiler_fence is equivalent to C++ [`atomic_signal_fence`](https://en.cppreference.com/w/cpp/atomic/atomic_signal_fence), and so its synchronization rules are very similar to those of `fence` except that synchronization only occurs between a thread and its signal handler, not across different threads.

Cc @rust-lang/libs-api @rust-lang/opsem 
Cc https://github.com/rust-lang/unsafe-code-guidelines/issues/347 https://github.com/rust-lang/rust/issues/54962

### PR DESCRIPTION
**Pull request by RalfJung**
Our `fence` docs have at some point been update to explain that they are about synchronization, not about "preventing reordering". This updates the `compiler_fence` docs n the same vein, mostly by referring to the `fence` docs.

The old docs make it sound like I can put a compiler_fence in the middle of a bunch of non-atomic operations and that would achieve any kind of guarantee. It does not, atomic operations are still required to do synchronization.

I also slightly tweaked the `fence` docs, to put the synchronization first and the "prevent reordering" second.

Cc @rust-lang/opsem @chorman0773 @m-ou-se 

Fixes https://github.com/rust-lang/rust/issues/129189
Fixes https://github.com/rust-lang/rust/issues/54962

### CODE REVIEW

**Comment by chorman0773** [comment]
Ideally, I'd remove all of the mentions of reordering from the docs.

Otherwise, looks to be correct.

**Comment by RalfJung** [comment]
> Ideally, I'd remove all of the mentions of reordering from the docs.

Yeah... I didn't dare suggest that as it is such a common thing to talk about in weak memory models. So I opted for just emphasizing them less.

File: library/core/src/sync/atomic.rs [hunk 1]
```diff
@@ -3659,33 +3664,30 @@ pub fn fence(order: Ordering) {
...
+///   as the code it interrupted, but it is not ordered with respect to that code. `compiler_fence`
+///   can be used to establish synchronization between a thread and its signal handler, the same way
+///   that `fence` can be used to establish synchronization across threads.
+/// - Similar situations can arise in embedded programming with interrupt handlers, or in custom
+///   implementations of preemptive green threads. In general, `compiler_fence` can establish
+///   synchronization with code that is guaranteed to run on the same hardware CPU.
 ///
-/// The re-ordering prevented by the different ordering semantics are:
+/// See [`fence`] for how a fence can be used to achieve synchronization. Note that just like
+/// [`fence`], synchronization still requires atomic operations to be used in both threads -- it is
```
Comment (thomcc) [inline library/core/src/sync/atomic.rs:3683]: TIL, I thought they could be volatile too.

**Review by thomcc — APPROVED** [review]
LGTM

File: library/core/src/sync/atomic.rs (hunk 1 above)
Comment (RalfJung) [reply library/core/src/sync/atomic.rs:3683]: Yeah would be nice to use this for volatile ordering, but that's not per-spec... and IMO the fix for that is to make volatile operations also be atomic, but that project has stalled it seems.

**Pull request merged**